ethcore/light/src/on_demand/request_guard.rs

```diff
@@ -0,0 +1,89 @@
+// Copyright 2015-2018 Parity Technologies (UK) Ltd.
+// This file is part of Parity.
+
+// Parity is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Parity is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Parity.  If not, see <http://www.gnu.org/licenses/>.
+
+use failsafe;
+use std::time::Duration;
+
+type RequestPolicy = failsafe::failure_policy::SuccessRateOverTimeWindow<failsafe::backoff::Exponential>;
+
+/// Error wrapped on-top of `FailsafeError`
+#[derive(Debug)]
+pub enum Error {
+	/// The call is let through
+	LetThrough,
+	/// The call rejected by the guard
+	Rejected,
+	/// The request reached the maximum of backoff iterations
+	ReachedLimit,
+}
+
+/// Handle and register requests that can fail
+#[derive(Debug)]
+pub struct RequestGuard {
+	num_failures: usize,
+	max_failures: usize,
```

Comment:
Can it be implemented as `success_rate_over_time_window(...).or_else(consecutive_failures(...))`?

Reply:
AFAIK, it can not because `failsafe-rs` will never drop a request it will just `backoff` and continue forever and we can't access the internal state. Thus, we can't distinguish between the `two failures`, i.e, if the request is in `Rejected state` just backed off or if it should be dropped. But, I will double check this!

However, this could also be implemented by a specific failure policy but this was pretty easy to implement so I didn't bother

Reply:
ok, makes sense